spec/queries/matched_for_notification_spec.rb

```diff
@@ -1,73 +1,177 @@
+        expect(lists).to eq([@subscriber_list3, @subscriber_list4, @and_joined_facet_subscriber_list3, @and_joined_facet_subscriber_list4])
 
-      lists = execute_query(field: :links, query_hash: { topics: ["uuid-123"] })
-      expect(lists).to eq([@list4])
-    end
+        lists = execute_query(field: :links, query_hash: { topics: ["uuid-123"] })
+        expect(lists).to eq([@subscriber_list4, @and_joined_facet_subscriber_list4])
+      end
+
+      it 'finds subscriber lists matching all topics' do
+        lists = execute_query(field: :links, query_hash: { topics: ["uuid-234", "uuid-123"] })
+        expect(lists).to include(@and_joined_facet_subscriber_list5)
+      end
 
-    it 'finds subscriber lists matching all topics' do
-      lists = execute_query(field: :links, query_hash: { topics: ["uuid-234", "uuid-123"] })
-      expect(lists).to include(@list5)
+      it 'finds subscriber lists matching any and all topics' do
+        lists = execute_query(field: :links, query_hash: { topics: ["uuid-345", "uuid-678", "uuid-456"] })
+        expect(lists).to include(@and_joined_facet_subscriber_list6)
+      end
+
+      it 'finds subscriber lists matching a mix of any and all topics and policies' do
+        lists = execute_query(field: :links, query_hash: { topics: ["uuid-345", "uuid-456", "other1"],
+                                                           policies: ["uuid-567", "uuid-678", "other2"] })
+        expect(lists).to include(@and_joined_facet_subscriber_list7)
+      end
     end
 
-    it 'finds subscriber lists matching any and all topics' do
-      lists = execute_query(field: :links, query_hash: { topics: ["uuid-345", "uuid-678", "uuid-456"] })
-      expect(lists).to include(@list6)
+    context "finds or joined subscription lists for the right query hash" do
+      it 'will match an OrJoinedSubscriberList with multiple facets where there is only one matching facet' do
+        lists = execute_query(field: :tags, query_hash: { topics: %w[topic_list_1] })
+        expect(lists).to eq([@or_joined_facet_subscriber_list1])
+      end
+
+      it 'will match an OrJoinedSubscriberList with multiple facets where all facets match' do
+        lists = execute_query(field: :tags, query_hash: { topics: %w[topic_list_1], organisations: %w[organisation_1_list_1 organisation_2_list_1] })
+        expect(lists).to eq([@or_joined_facet_subscriber_list1])
+      end
+
+      it 'will not match an OrJoinedSubscriberList with multiple facets where no facets match' do
+        lists = execute_query(field: :tags, query_hash: { format: %w[employment_tribunal_decision] })
+        expect(lists).to eq([])
+      end
+
+      it 'will not match an OrJoinedSubscriberList if the query match only matches one of just one facet that has an all' do
+        lists = execute_query(field: :tags, query_hash: { organisations: %w[organisation_1_list_2] })
```

Comment:
Think these tests would be easier to read if they followed the Arrange, Act, Assert model. Currently the 'Arrange' all happens at the top and you have to scroll up and down to figure out what's going on.

I also wonder if a helper method might make the 'Arrange' parts easier to read.

That way we can do away with the naming that doesn't really mean anything ("organisation 1 list 2")

Reply:
Agreed, this was hard to read when I started and now it's worse. Will have a think about ways to improve it